File: kgk_customisations/finance_management/report/bank_balance_report/bank_balance_report.js
// Copyright (c) 2026, Apjakal IT Solutions and contributors
// For license information, please see license.txt

frappe.query_reports["Bank Balance Report"] = {
	filters: [
		{
			fieldname: "date_from",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			reqd: 1,
		},
		{
			fieldname: "date_to",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
	],

	onload: function (report) {
		const _orig = report.render_datatable.bind(report);
		report.render_datatable = function () {
			report.datatable = null;
			_orig();
		};
	},

	after_datatable_render: function (datatable_obj) {
		const data    = frappe.query_report.data    || [];
		const columns = frappe.query_report.columns || [];

		const banks = [];
		columns.forEach((col) => {
			const m = (col.fieldname || "").match(/^bank_(\d+)_accountant$/);
			if (m) banks.push({ idx: m[1], label: (col.label || "").replace(/ Lore$/, "") });
		});

		const fmt = (v) =>
			frappe.format(v || 0, { fieldtype: "Currency" }, { always_show_decimals: true });

		let thead = `<thead><tr style="background:#f8f9fa">
			<th rowspan="2" style="${bbr_th()}text-align:left;min-width:100px">Date</th>`;
		banks.forEach((b) => {
			thead += `<th colspan="2" style="${bbr_th()}text-align:center">${b.label}</th>`;
		});
		thead += `</tr><tr style="background:#f8f9fa">`;
		banks.forEach(() => {
			thead += `<th style="${bbr_th()}text-align:right">Lore</th>
			          <th style="${bbr_th()}text-align:right">Harsh</th>`;
		});
		thead += `</tr></thead>`;

		let tbody = "<tbody>";
		data.forEach((row, ri) => {
			const rowBg = ri % 2 === 0 ? "#fff" : "#f8f9fa";
			tbody += `<tr style="background:${rowBg}">
				<td style="${bbr_td()}font-weight:500">${row.date}</td>`;
			banks.forEach((b) => {
				const acct    = row[`bank_${b.idx}_accountant`] || 0;
				const checker = row[`bank_${b.idx}_checker`]    || 0;
				const cellBg  = row[`tally_${b.idx}`] ? "#d4f0dc" : "#fde8e8";
				tbody += `<td style="${bbr_td()}text-align:right;background:${cellBg}">${fmt(acct)}</td>
				          <td style="${bbr_td()}text-align:right;background:${cellBg}">${fmt(checker)}</td>`;
			});
			tbody += `</tr>`;
		});
		tbody += `</tbody>`;

		$(datatable_obj.wrapper).html(`
			<div style="overflow-x:auto;padding:12px 4px">
				<table style="border-collapse:collapse;width:100%;font-size:13px;font-family:sans-serif">
					${thead}${tbody}
				</table>
			</div>`);
	},
};

function bbr_th() { return "padding:10px 12px;color:#800000;background:#f8f9fa;border-bottom:2px solid #d1d8dd;white-space:nowrap;"; }
function bbr_td() { return "padding:8px 12px;color:#444;border-bottom:1px solid #ebeff2;white-space:nowrap;"; }

File: kgk_customisations/finance_management/report/bank_balance_report/bank_balance_report.py
# Copyright (c) 2026, Apjakal IT Solutions and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt
from kgk_customisations.finance_management.doctype.cash_balance.cash_balance import _load_aggregates

_BANKS = ["ZAR@ABSA", "USD@ABSA", "BWP@FNB", "BWP@ABSA"]
_ACCOUNT_NUMBERS = {
	"ZAR@ABSA": "1008748",
	"USD@ABSA": "1004556",
	"BWP@FNB": "62415220110",
	"BWP@ABSA": "1023469",
}

COMPANY_MAP = {
				"Diamonds": "KGK Diamonds",
				"Jewellery": "KGK Jewelry",
				"Agro": "KGK Agro",
				"Healthcare": "KGK Healthcare",
}


def execute(filters=None):
	if not filters:
		filters = {}
	date_from = filters.get("date_from")
	date_to   = filters.get("date_to")
	if date_from and date_to:
		banks = _discover_banks(date_from, date_to)
	else:
		banks = list(_BANKS)
	columns = _get_columns(banks)
	data    = _get_data(filters, banks)
	return columns, data


def _discover_banks(date_from, date_to):
	"""Return an ordered list of unique bank keys present in the date range.

	Keys from Cash Balance (balance_type=Bank) and Bank Balance Entry.account
	are merged. Known banks (_BANKS) appear first in declared order; additional
	keys are appended alphabetically.
	"""
	cb_rows = frappe.db.sql(
		"SELECT DISTINCT company FROM `tabCash Balance`"
		" WHERE balance_type = 'Bank' AND date BETWEEN %s AND %s",
		[date_from, date_to],
	)
	cb_keys = {str(row[0]) for row in cb_rows if row[0]}

	bb_rows = frappe.db.sql(
		"SELECT DISTINCT account FROM `tabBank Balance Entry`"
		" WHERE account IS NOT NULL AND account != ''"
		" AND date BETWEEN %s AND %s",
		[date_from, date_to],
	)
	bb_keys = {str(row[0]) for row in bb_rows if row[0]}

	all_keys = cb_keys | bb_keys
	ordered = [b for b in _BANKS if b in all_keys]
	extras  = sorted(all_keys - set(_BANKS))
	return ordered + extras


def _get_columns(banks):
	columns = [
		{
			"fieldname": "date",
			"label": "Date",
			"fieldtype": "Date",
			"width": 100,
		}
	]
	for i, bank in enumerate(banks):
		acc_no = _ACCOUNT_NUMBERS.get(bank)
		label = "{} ({})".format(bank, acc_no) if acc_no else bank
		columns += [
			{
				"fieldname": "bank_{}_accountant".format(i),
				"label": "{} Lore".format(label),
				"fieldtype": "Currency",
				"width": 150,
			},
			{
				"fieldname": "bank_{}_checker".format(i),
				"label": "{} Harsh".format(label),
				"fieldtype": "Currency",
				"width": 150,
			},
		]
	# Hidden tally columns used by JS formatter (1=match, 0=mismatch)
	for i in range(len(banks)):
		columns.append(
			{
				"fieldname": "tally_{}".format(i),
				"label": "Tally {}".format(i),
				"fieldtype": "Int",
				"hidden": 1,
			}
		)
	return columns


def _get_data(filters, banks):
	date_from = filters.get("date_from")
	date_to = filters.get("date_to")
	if not date_from or not date_to:
		return []

	# Cash Checker (without Super User) sees only their own BankBasicEntry rows
	user_roles = frappe.get_roles(frappe.session.user)
	is_checker = (
		"Cash Checker" in user_roles
		and "Cash Super User" not in user_roles
		and frappe.session.user != "Administrator"
	)
	current_user = frappe.session.user

	# Collect all relevant dates from both tables
	agg = _load_aggregates(date_from=date_from, date_to=date_to, balance_type="Bank")
	cb_dates = [(d,) for (d, bt, key), vals in agg.items() if bt == "Bank"]
	bb_dates = frappe.db.sql(
		"SELECT DISTINCT date FROM `tabBank Balance Entry`"
		" WHERE account IS NOT NULL AND account != ''"
		" AND date BETWEEN %s AND %s",
		[date_from, date_to],
	)
	all_dates = sorted(
		set(str(row[0]) for row in cb_dates) | set(str(row[0]) for row in bb_dates)
	)

	data = []
	for date in all_dates:
		date_str = str(date)

		# Accountant values: Cash Balance (balance_type=Bank), company = bank key
		acct_rows = frappe.db.get_all(
			"Cash Balance",
			filters={"date": date_str, "balance_type": "Bank"},
			fields=["company", "accountant"],
		)
		acct_map = {r.company: flt(r.accountant) for r in acct_rows}

		# Checker values from Bank Balance Entry (account field)
		if is_checker:
			bb_rows = frappe.db.get_all(
				"Bank Balance Entry",
				filters={"date": date_str, "username": current_user},
				fields=["account", "balance"],
			)
			checker_map = {r.account: flt(r.balance) for r in bb_rows if r.account}
		else:
			bb_rows = frappe.db.sql(
				"SELECT account, SUM(balance) AS total"
				" FROM `tabBank Balance Entry`"
				" WHERE date = %s AND account IS NOT NULL AND account != ''"
				" GROUP BY account",
				[date_str],
				as_dict=True,
			)
			checker_map = {r.account: flt(r.total) for r in bb_rows}

		row = {"date": date_str}
		nonzero = False

		for i, bank in enumerate(banks):
			accountant = acct_map.get(bank, 0.0)
			checker = checker_map.get(bank, 0.0)
			tally = 1 if abs(accountant - checker) < 1e-6 else 0

			row["bank_{}_accountant".format(i)] = accountant
			row["bank_{}_checker".format(i)] = checker
			row["tally_{}".format(i)] = tally

			if abs(accountant) > 1e-6 or abs(checker) > 1e-6:
				nonzero = True

		if nonzero:
			data.append(row)

	return data
